test/test-parse-warning.cpp

```diff
@@ -672,6 +672,97 @@ TEST_F(test_parse_warning,
         }));
   }
 }
+
+
+TEST_F(test_parse_warning,
+       warn_on_pointless_nullish_coalescing_operator) {
+  {
+    test_parser p(u8"true ?? false"_sv, capture_diags);
+    p.parse_and_visit_expression();
+
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"true "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"(a < b) ?? false"_sv, capture_diags);
+    p.parse_and_visit_expression();
+
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"(a < b) "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"!b ?? false"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"!b "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"'hi' ?? true"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"'hi' "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"s.toLowerCase() ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"s ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"null ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"(foo) ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"{}.missingProp ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"{}['missingProp'] ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"await foo ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"void 42 ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"bar`hello` ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
+  {
+    test_parser p(u8"this ?? false"_sv);
+    p.parse_and_visit_expression();
+  }
```

Comment:
Nitpick: To shrink this code further, we could use a loop:
```suggestion
  for (string8_view code : {
           u8"s.toLowerCase() ?? false"_sv,
           u8"s ?? false"_sv,
           u8"null ?? false"_sv,
           u8"(foo) ?? false"_sv,
           u8"{}.missingProp ?? false"_sv,
           u8"{}['missingProp'] ?? false"_sv,
           u8"await foo ?? false"_sv,
           u8"void 42 ?? false"_sv,
           u8"bar`hello` ?? false"_sv,
           u8"this ?? false"_sv,
       }) {
    SCOPED_TRACE(out_string8(code));
    test_parser p(code);
    p.parse_and_visit_expression();
  }
```